File: web/src/lib/states.ts
import type { LeaseState } from "@/lib/types";

/**
 * How a lease state looks and what it means, mirrored from `bailment.states`.
 *
 * **Colour is a claim, so it is rationed.** Green means one thing here: a resource exists
 * and you may use it right now. `released` is the successful end of a lease and it is
 * still grey, because a table of fifty finished leases rendered in green drowns the two
 * that are live -- and "which of these can I use" is the question the table is for.
 *
 * **Orphaned is amber, never red.** An orphan is the finding this project exists to
 * produce. Rendering it as an alarm teaches operators that the reconciler is a thing that
 * goes wrong, when it is the thing that is working; and an interface that shouts is an
 * interface people learn to dismiss. Amber says "this needs a decision", which is exactly
 * true. `failed` is the only red, and only because it is terminal and nothing further will
 * happen without a person.
 *
 * The blurbs are the state machine's own docstrings, condensed. They appear on hover, so
 * somebody meeting `deprovisioning` for the first time does not have to go and read
 * states.py.
 */

export type Tone = "neutral" | "info" | "ok" | "warn" | "danger";

interface StateMeta {
  label: string;
  tone: Tone;
  blurb: string;
}

const META: Record<LeaseState, StateMeta> = {
  pending: {
    label: "Pending",
    tone: "neutral",
    blurb: "Request accepted and persisted. No policy decision made yet.",
  },
  awaiting_approval: {
    label: "Awaiting approval",
    tone: "info",
    blurb: "Policy said a human must decide. Nothing has been provisioned.",
  },
  rejected: {
    label: "Rejected",
    tone: "neutral",
    blurb: "Policy denied it, or an approver declined. Terminal, and nothing was created.",
  },
  provisioning: {
    label: "Provisioning",
    tone: "info",
    blurb: "A worker holds the claim and is calling the provider.",
  },
  active: {
    label: "Active",
    tone: "ok",
    blurb: "The provider confirmed the resource exists. The lease clock is running.",
  },
  expiring: {
    label: "Expiring",
    tone: "warn",
    blurb: "Inside the warning window. Still fully usable; a notice has been emitted.",
  },
  expired: {
    label: "Expired",
    tone: "neutral",
    blurb: "The TTL elapsed. An intention to destroy — the resource may still exist.",
  },
  revoked: {
    label: "Revoked",
    tone: "neutral",
    blurb: "Ended early by a human or by policy. Also an intention, not a fact.",
  },
  deprovisioning: {
    label: "Deprovisioning",
    tone: "info",
    blurb: "A worker holds the claim and is calling the provider's delete path.",
  },
  released: {
    label: "Released",
    tone: "neutral",
    blurb: "The provider confirmed the resource is gone. The only clean terminal state.",
  },
  failed: {
    label: "Failed",
    tone: "danger",
    blurb: "Provisioning failed and any partial resource was rolled back. Terminal.",
  },
  orphaned: {
    label: "Orphaned",
    tone: "warn",
    blurb:
      "A real resource is believed to exist that no live lease accounts for. This is the " +
      "state the whole project exists to make visible.",
  },
  unknown: {
    label: "Unknown",
    tone: "warn",
    blurb: "Provider state could not be determined. Never assumed released; always re-checked.",
  },
};

const FALLBACK: StateMeta = {
  label: "Unrecognised",
  tone: "neutral",
  blurb: "This broker reported a lease state this dashboard does not know about.",
};

export function stateMeta(state: string): StateMeta {
  return META[state as LeaseState] ?? { ...FALLBACK, label: state };
}

export function stateLabel(state: string): string {
  return stateMeta(state).label;
}

export function stateTone(state: string): Tone {
  return stateMeta(state).tone;
}

/**
 * The order the filter control offers states in: lifecycle order, matching the enum in
 * states.py, because reading it top to bottom should tell you the story.
 */
export const STATE_FILTER_GROUPS: Array<{ label: string; states: LeaseState[] }> = [
  { label: "Intake", states: ["pending", "awaiting_approval", "rejected"] },
  { label: "Live", states: ["provisioning", "active", "expiring"] },
  { label: "Teardown", states: ["expired", "revoked", "deprovisioning", "released"] },
  { label: "Trouble", states: ["failed", "orphaned", "unknown"] },
];

/** Tailwind classes for a badge in each tone. One definition, used everywhere. */
export const TONE_BADGE: Record<Tone, string> = {
  neutral: "bg-muted text-muted-foreground ring-border",
  info: "bg-info-soft text-info ring-info/20",
  ok: "bg-ok-soft text-ok ring-ok/20",
  warn: "bg-warn-soft text-warn ring-warn/25",
  danger: "bg-danger-soft text-danger ring-danger/20",
};

/** The dot that precedes a badge label. Solid, so it survives being seen out of focus. */
export const TONE_DOT: Record<Tone, string> = {
  neutral: "bg-muted-foreground/60",
  info: "bg-info",
  ok: "bg-ok",
  warn: "bg-warn",
  danger: "bg-danger",
};

File: src/bailment/states.py
"""The lease lifecycle state machine.

This module is the single authority on which state transitions are legal. Everything
else in bailment -- the API, the workers, the reconciler, the dashboard -- reads the
transition table from here rather than encoding its own idea of the lifecycle.

The design rule that matters: a resource is only ever considered gone once the
*provider* has confirmed it is gone. A row moving to ``RELEASED`` is a claim about
reality, so nothing may write that state except the deprovision path and the
reconciler. Everything else that wants a lease to end asks for ``EXPIRED`` or
``REVOKED``, which are *intentions*, and lets the engine do the work.

That separation is what makes orphan detection possible at all: if code could write
``RELEASED`` optimistically, a failed destroy would look identical to a successful one
and the orphan would be invisible forever.
"""

from __future__ import annotations

from enum import StrEnum


class LeaseState(StrEnum):
    """Every state a request/lease can occupy.

    Ordered roughly by lifecycle position, not alphabetically, because reading this
    enum top to bottom should tell you the story.
    """

    # --- Intake -------------------------------------------------------------
    PENDING = "pending"
    """Request accepted and persisted. No policy decision made yet."""

    AWAITING_APPROVAL = "awaiting_approval"
    """Policy said a human must decide. Nothing has been provisioned."""

    REJECTED = "rejected"
    """Policy denied it outright, or an approver declined. Terminal."""

    # --- Provisioning -------------------------------------------------------
    PROVISIONING = "provisioning"
    """A worker holds the claim and is calling the provider. Crash-prone window."""

    ACTIVE = "active"
    """Provider confirmed the resource exists. The lease clock is running."""

    EXPIRING = "expiring"
    """Inside the warning window. Still fully usable; a notice has been emitted."""

    # --- Teardown -----------------------------------------------------------
    EXPIRED = "expired"
    """TTL elapsed. An *intention* to destroy -- the resource may still exist."""

    REVOKED = "revoked"
    """A human or policy ended it early. Also an intention, not a fact."""

    DEPROVISIONING = "deprovisioning"
    """A worker holds the claim and is calling the provider's delete path."""

    RELEASED = "released"
    """Provider confirmed the resource is gone. The only clean terminal state."""

    # --- Failure modes ------------------------------------------------------
    FAILED = "failed"
    """Provisioning failed and any partial resource was rolled back. Terminal."""

    ORPHANED = "orphaned"
    """We believe a real resource exists that no live lease accounts for.

    Reached either when a destroy fails repeatedly, or when the reconciler finds a
    resource at the provider with no matching active lease. This is the state the
    whole project exists to make visible.
    """

    UNKNOWN = "unknown"
    """Provider state could not be determined. Never assume; always re-check."""


#: States where a real resource is believed to exist and cost money.
LIVE_STATES: frozenset[LeaseState] = frozenset(
    {
        LeaseState.PROVISIONING,
        LeaseState.ACTIVE,
        LeaseState.EXPIRING,
        LeaseState.EXPIRED,
        LeaseState.REVOKED,
        LeaseState.DEPROVISIONING,
        LeaseState.ORPHANED,
        LeaseState.UNKNOWN,
    }
)

#: States a consumer may actually use the binding from.
USABLE_STATES: frozenset[LeaseState] = frozenset({LeaseState.ACTIVE, LeaseState.EXPIRING})

#: No further transition is possible. The reconciler skips these.
TERMINAL_STATES: frozenset[LeaseState] = frozenset(
    {LeaseState.REJECTED, LeaseState.RELEASED, LeaseState.FAILED}
)

#: States that mean "someone decided this should end" but teardown has not finished.
#: The lease engine picks these up and drives them to DEPROVISIONING.
TEARDOWN_INTENT_STATES: frozenset[LeaseState] = frozenset({LeaseState.EXPIRED, LeaseState.REVOKED})

#: The complete legal transition table. Anything absent here is a bug, not a feature.
TRANSITIONS: dict[LeaseState, frozenset[LeaseState]] = {
    LeaseState.PENDING: frozenset(
        {
            LeaseState.AWAITING_APPROVAL,
            LeaseState.PROVISIONING,
            LeaseState.REJECTED,
            LeaseState.FAILED,
        }
    ),
    LeaseState.AWAITING_APPROVAL: frozenset(
        {
            LeaseState.PROVISIONING,
            LeaseState.REJECTED,
            # An approval request can outlive its own deadline.
            LeaseState.EXPIRED,
        }
    ),
    LeaseState.PROVISIONING: frozenset(
        {
            LeaseState.ACTIVE,
            LeaseState.FAILED,
            # Provider call succeeded but we could not read back the result.
            LeaseState.UNKNOWN,
            # Revoked mid-flight; teardown must still run.
            LeaseState.REVOKED,
        }
    ),
    LeaseState.ACTIVE: frozenset(
        {
            LeaseState.EXPIRING,
            LeaseState.EXPIRED,
            LeaseState.REVOKED,
            LeaseState.UNKNOWN,
        }
    ),
    LeaseState.EXPIRING: frozenset(
        {
            LeaseState.EXPIRED,
            LeaseState.REVOKED,
            # Extension granted before the deadline.
            LeaseState.ACTIVE,
            LeaseState.UNKNOWN,
        }
    ),
    LeaseState.EXPIRED: frozenset({LeaseState.DEPROVISIONING, LeaseState.UNKNOWN}),
    LeaseState.REVOKED: frozenset({LeaseState.DEPROVISIONING, LeaseState.UNKNOWN}),
    LeaseState.DEPROVISIONING: frozenset(
        {
            LeaseState.RELEASED,
            # Destroy failed. Not terminal -- it is now our problem to surface.
            LeaseState.ORPHANED,
            LeaseState.UNKNOWN,
        }
    ),
    LeaseState.ORPHANED: frozenset(
        {
            # Retry the destroy.
            LeaseState.DEPROVISIONING,
            # Reconciler confirmed it really is gone after all.
            LeaseState.RELEASED,
            LeaseState.UNKNOWN,
        }
    ),
    LeaseState.UNKNOWN: frozenset(
        {
            LeaseState.ACTIVE,
            LeaseState.ORPHANED,
            LeaseState.RELEASED,
            LeaseState.DEPROVISIONING,
            LeaseState.FAILED,
        }
    ),
    LeaseState.REJECTED: frozenset(),
    LeaseState.RELEASED: frozenset(),
    LeaseState.FAILED: frozenset(),
}


class IllegalTransition(Exception):
    """Raised when code attempts a transition the lifecycle does not permit."""

    def __init__(self, source: LeaseState, target: LeaseState) -> None:
        self.source = source
        self.target = target
        super().__init__(
            f"illegal lease transition {source.value} -> {target.value}; "
            f"legal targets are {sorted(t.value for t in TRANSITIONS[source]) or ['<terminal>']}"
        )


def can_transition(source: LeaseState, target: LeaseState) -> bool:
    """Return whether ``source -> target`` is a legal move."""
    return target in TRANSITIONS[source]


def assert_transition(source: LeaseState, target: LeaseState) -> None:
    """Raise :class:`IllegalTransition` unless ``source -> target`` is legal.

    Self-transitions are permitted and treated as no-ops so that idempotent retries
    of the same worker step do not blow up.
    """
    if source == target:
        return
    if not can_transition(source, target):
        raise IllegalTransition(source, target)


def is_live(state: LeaseState) -> bool:
    """Whether a real, probably billable resource is believed to exist."""
    return state in LIVE_STATES


def is_terminal(state: LeaseState) -> bool:
    """Whether the lease is finished and needs no further work."""
    return state in TERMINAL_STATES
